Add night light brightness level setting to VeSync

iprak opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR introduces night light brightness level setting which can be set on VeSync Classic300S humidifier physically via  touch button.

Classic300S supports 3 brightness leves: 0, 50 and 100. This was determined by changing the level physically on the humidifier and examining the debug status. The set operation was physically cross checked on the humidifer.

The setting is limited to device type Classic300S since that is what I could use to physically verify the end result.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/37332
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

iprak committed:
Add night light brightness level setting

Classic300S supports 3 brightness leves: 0, 50 and 100. This was
determined by changing the level physically on the humidifier and
examining the debug status. The set operation was physically
cross checked on the humidifer.

The setting is limited to device type VeSyncHumid200300S which
represents Classic300S.

cdnninja commented:
Should this be a number or a light entity?  Mine has a night light as well just off / on.

cdnninja commented on homeassistant/components/vesync/number.py:
```diff
@@ -44,7 +45,18 @@ class VeSyncNumberEntityDescription(NumberEntityDescription):
+        native_max_value=100,
+        native_step=50,
+        mode=NumberMode.SLIDER,
+        exists_fn=lambda device: isinstance(device, VeSyncHumid200300S),
```
Could we check for the data instead of just supporting this one device?

cdnninja commented on homeassistant/components/vesync/strings.json:
```diff
@@ -54,6 +54,9 @@
         "name": "Mist level"
+      },
+      "night_light_brightness": {
+        "name": "Night brightness level"
```
I think it should have the word light in it.

iprak replied:
I had picked this because the card only is only able to show part of that string.

![image](https://github.com/user-attachments/assets/abc09ca0-36f2-49e9-98a3-f712d3e4b08f)

I think "Nightlight level" feels the best.

frenck commented:
I've marked this PR as a draft, as changes are requested that need to be processed.
Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

iprak commented:
> Should this be a number or a light entity? Mine has a night light as well just off / on.

What model is this? I see a tri-level effect on my humidifier identified as Classic300S. The on/off button is for turning the humidifier itself.

cdnninja commented:
I stand corrected, I assumed mine was on off as the app only seems to show that.  Front panel does the multi levels. Mine is also a Classic300S.

iprak commented:
> I stand corrected, I assumed mine was on off as the app only seems to show that. Front panel does the multi levels. Mine is also a Classic300S.

Ok. I do not think this should be a light entity since at least Classic300s has 3 levels (0, 50, 100). It does not seem suitable as dimmer levels. Plus as we discovered the on/off is for the humidifier and not the light.

iprak committed:
Revert "Add night light brightness level setting"

This reverts commit 29e5a945dacaa2bc26a425c34d1060e3f0053743.

iprak committed:
Support night light level as select option.

Classic300S supports 3 brightness leves: 0, 50 and 100. This was
determined by changing the level physically on the humidifier and
examining the debug status. The set operation was physically
cross checked on the humidifer.

The setting is exposed as select entity with values off, dim and bright.

The setting is limited to devices supporting `night_light`.

iprak committed:
Add unit tests for the select platform.

iprak commented:
I reworked this as a select entity which feels better.

![image](https://github.com/user-attachments/assets/18e3c253-c3f0-4ed7-90ca-1796adcfe02e)


I reworked the entity to be available when device supports `night_light` property, the same property is internally checked when setting `set_night_light_brightness`.

abmantis commented on tests/components/vesync/test_select.py:
```diff
@@ -0,0 +1,43 @@
+from .common import ENTITY_HUMIDIFIER_300S_NIGHT_LIGHT_SELECT
+
+
+async def test_set_nightlight_level(
```
```suggestion
async def test_night_light_level(
```

can we also test reading the current value?

iprak replied:
The device in the test is a mocked device so it itself would not update. The `select` entity update could however be a part of test_platform but it currently does not handle humidifier device at all.

abmantis replied:
What I mean is that you can use the mocked device to set a state and ensure that the select state reflects it. The other tests for vesync do that (check `test_mist_level` in `test_number.py`, for example)

iprak committed:
vesync(select): code fixes

iprak committed:
vesync(test_select): verify value in api call

iprak committed:
select: pass in the correct param to async_add_entities

iprak committed:
select(test): verify entity state

abmantis commented on tests/components/vesync/test_select.py:
```diff
@@ -0,0 +1,73 @@
+        )
+
+        # Assert that setter API was invoked with the expected translated value
+        method_mock.assert_called_once_with(
```
```suggestion
    humidifier_300s.set_night_light_brightness.assert_called_once_with(
```
no need for the `with patch` on L36

abmantis commented on tests/components/vesync/test_select.py:
```diff
@@ -0,0 +1,73 @@
+            return_value=[humidifier_300s],
+        ),
+        patch(
+            "homeassistant.helpers.update_coordinator.DataUpdateCoordinator.async_request_refresh"
```
I think we can avoid patching the internals and patch the vesync api instead here. maybe `VeSync.update_all_devices`?

iprak replied:
The Vesync manager is already being mocked so the mocked device can simply be stuffed into the device list.

iprak replied:
Changes not ready yet. I will address it tonight.

abmantis commented on tests/components/vesync/test_select.py:
```diff
@@ -0,0 +1,73 @@
+) -> None:
+    """Test the state of night light level select entity."""
+
+    # The mocked device has night_light_brightness=50 which is "dim"
```
lets move this to L70 since that is where the check is happening

iprak committed:
select(test): simplify and correct tests

* Consolidated humidifier device setup into a fixture. Vesync was
already being mocked so we can just install the humidifier as a fan.
* manager.update_all_devices is asserted instead of patching
DataUpdateCoordinator.async_request_refresh

abmantis reviewed (approved):
Thanks!
